parsl/dataflow/dflow.py

```diff
@@ -352,21 +355,36 @@ def handle_exec_update(self, task_record, future):
                     # record the inner app ID in monitoring, and add a completion
                     # listener to that inner future.
 
-                    inner_future = future.result()
+                    joinable = future.result()
 
                     # Fail with a TypeError if the joinapp python body returned
                     # something we can't join on.
-                    if isinstance(inner_future, Future):
+                    if isinstance(joinable, Future):
+                        task_record['status'] = States.joining
+                        task_record['joins'] = joinable
+                        task_record['join_lock'] = threading.Lock()
+                        joinable.add_done_callback(partial(self.handle_join_update, task_record))
+                    elif isinstance(joinable, list):  # TODO: should this be list or arbitrary iterable?
                         task_record['status'] = States.joining
-                        task_record['joins'] = inner_future
-                        inner_future.add_done_callback(partial(self.handle_join_update, task_record))
+                        task_record['joins'] = joinable
+                        task_record['join_lock'] = threading.Lock()
+                        for inner_future in joinable:
+                            # TODO: typechecking and error setting here - perhaps
+                            # should put this and the one-future case inside a
+                            # try and perform the error handling there in an
+                            # except block? (it would be ok to go joining->failed
+                            # which doesn't happen in the type error case but
+                            # does happen in the joined-tasks fail case)
+                            # For now, this assert will cause a DFK hang
```

Comment:
Could we trigger a shutdown if the assert fails? Better to fail loudly than hang.

Reply:
probably i should figure out proper handling here